posthog/hogql/transforms/projection_pushdown.py

```diff
@@ -0,0 +1,170 @@
+        self.subquery_map: dict[int, ast.SelectQuery] = {}
+
+    def visit_select_query(self, node: ast.SelectQuery) -> ast.SelectQuery:
+        # Phase 1: Register subqueries for demand tracking
+        if node.select_from:
+            self._register_subqueries(node.select_from)
+
+        # Phase 2: Collect column demands from query clauses
+        # Skip asterisk-expanded SELECT columns - their demands flow from parent queries
+        for expr in node.select:
+            if not self._is_from_asterisk(expr):
+                self.visit(expr)
+
+        if node.where:
+            self.visit(node.where)
+        if node.prewhere:
+            self.visit(node.prewhere)
+        if node.group_by:
+            for expr in node.group_by:
+                self.visit(expr)
+        if node.having:
+            self.visit(node.having)
+        if node.order_by:
+            for expr in node.order_by:
+                self.visit(expr)
+        if node.limit:
+            self.visit(node.limit)
+        if node.offset:
+            self.visit(node.offset)
+
+        if node.select_from:
+            self._collect_join_constraint_column_demands(node.select_from)
+
+        # Phase 3: Propagate parent demands down to child subqueries
+        self._propagate_demands_to_children(node)
+
+        # Phase 4: Recursively visit and optimize child subqueries
+        if node.select_from:
+            self.visit(node.select_from)
+
+        # Phase 5: Prune unreferenced asterisk columns from this query
+        self._prune_columns(node)
+
+        return node
+
+    def _is_from_asterisk(self, expr: ast.Expr) -> bool:
+        """Check if an expression was expanded from asterisk"""
+        return isinstance(expr, ast.Field | ast.Alias) and getattr(expr, "from_asterisk", False)
+
+    def _propagate_demands_to_children(self, node: ast.SelectQuery) -> None:
+        """
+        Propagate parent demands to child subqueries.
+
+        When a parent query demands a column from us, we need to visit that column's
+        source expression to propagate the demand down to our child subqueries.
+        """
+        demanded_from_this = self.demands.get(id(node))
+        if not demanded_from_this:
+            return
+
+        for col_name in demanded_from_this:
+            for expr in node.select:
+                if self._get_column_name(expr) == col_name:
+                    self.visit(expr)
+                    break
+
+    def _register_subqueries(self, from_clause: ast.JoinExpr) -> None:
+        """Register all subqueries in FROM clause before collecting demands"""
+        if isinstance(from_clause.table, ast.SelectQuery):
+            self._register_subquery(from_clause.table, from_clause.type)
+
+        if from_clause.next_join:
+            self._register_subqueries(from_clause.next_join)
+
+    def _register_subquery(self, subquery: ast.SelectQuery, type_annotation: ast.Type | None) -> None:
```

Comment:
Is it possible for `type_annotation` to be `None`? We're running this after type resolution, right? So I would expect everything to have a type